File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    let first = -1;
    let start = 0;
    let end = nums.length - 1;

    while (start <= end) {
        const mid = Math.floor((start + end) / 2);
        if (nums[mid] === target) {
            first = mid;
            end = mid - 1;
        } else if (target < nums[mid]) {
            end = mid - 1
        } else {
            start = mid + 1;
        } 
    }

    let last = -1
    start = 0
    end = nums.length - 1
    while (start <= end) {
        const mid = Math.floor((start + end) /2);
        if (nums[mid] === target){
            last = mid
            start = mid + 1;
        } else if (target < nums[mid]) {
            end = mid - 1;
        } else {
            start = mid + 1;
        }
    }

    return [first, last]
};

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        targetIdx = -1

        start = 0
        end = len(nums) - 1

        while start <= end:
            mid = math.floor((start + end) /2)
            if (nums[mid] == target):
                targetIdx = mid
                break
            elif target < nums[mid]:
                end = mid - 1
            else:
                start = mid + 1

        if targetIdx == -1: return [-1, -1]

        start = end = targetIdx
        first = last = targetIdx

        while (start >= 0 and nums[start] == target):
            if nums[start] == target: first = start
            start -= 1
        
        while (end <= len(nums) - 1 and nums[end] == target):
            if (nums[end] == target): last = end
            end += 1
        return [first, last]